api/src/main/java/io/druid/data/input/impl/prefetch/PrefetchableTextFilesFirehoseFactory.java

```diff
@@ -0,0 +1,248 @@
+ * objects.
+ * <br/>
+ * - Retry: if an exception occurs while downloading an object, it retries again up to {@link #maxFetchRetry}.
+ * <p/>
+ *
+ * This implementation can be useful when the cost for reading input objects is large as reading from AWS S3 because
+ * batch tasks like IndexTask or HadoopIndexTask can read the whole data twice for determining partition specs and
+ * generating segments if the intervals of GranularitySpec is not specified.
+ * <br/>
+ * Prefetching can be turned on/off by setting maxFetchCapacityBytes.  Depending on prefetching is enabled or
+ * disabled, the behavior of the firehose is different like below.
+ * <p/>
+ *
+ * 1. If prefetch is enabled, this firehose can fetch input objects in background.
+ * <br/>
+ * 2. When next() is called, it first checks that there are already fetched files in local storage.
+ * <br/>
+ *   2.1 If exists, it simply chooses a fetched file and returns a {@link LineIterator} reading that file.
+ *   <br/>
+ *   2.2 If there is no fetched files in local storage but some objects are still remained to be read, the firehose
+ *   fetches one of input objects in background immediately. If an IOException occurs while downloading the object,
+ *   it retries up to the maximum retry count. Finally, the firehose returns a {@link LineIterator} only when the
+ *   download operation is successfully finished.
+ *   <br/>
+ * 3. If prefetch is disabled, the firehose returns a {@link LineIterator} which directly reads the stream opened by
+ * {@link #openObjectStream}. If there is an IOException, it will throw it and the read will fail.
+ */
+public abstract class PrefetchableTextFilesFirehoseFactory<ObjectType>
+    extends AbstractTextFilesFirehoseFactory<ObjectType>
+{
+  private static final Logger LOG = new Logger(PrefetchableTextFilesFirehoseFactory.class);
+  private static final long DEFAULT_MAX_CACHE_CAPACITY_BYTES = 1024 * 1024 * 1024; // 1GB
+  private static final long DEFAULT_MAX_FETCH_CAPACITY_BYTES = 1024 * 1024 * 1024; // 1GB
+  private static final long DEFAULT_FETCH_TIMEOUT = 60_000; // 60 secs
+  private static final int DEFAULT_MAX_FETCH_RETRY = 3;
+
+  private final CacheManager<ObjectType> cacheManager;
+  private final long maxFetchCapacityBytes;
+  private final long prefetchTriggerBytes;
+  private final long fetchTimeout;
+  private final int maxFetchRetry;
+
+  private List<ObjectType> objects;
+
+  private static ExecutorService createFetchExecutor()
+  {
+    return Executors.newSingleThreadExecutor(
+        new ThreadFactoryBuilder()
+            .setNameFormat("firehose_fetch_%d")
+            .build()
+    );
+  }
+
+  public PrefetchableTextFilesFirehoseFactory(
+      Long maxCacheCapacityBytes,
+      Long maxFetchCapacityBytes,
+      Long prefetchTriggerBytes,
+      Long fetchTimeout,
+      Integer maxFetchRetry
+  )
+  {
+    this.cacheManager = new CacheManager<>(maxCacheCapacityBytes == null
```

Comment:
Please extract argument as a variable

Reply:
I think this is better because the extracted variable looks redundant to me. BTW, fixed formatting.